### DISCUSSION TITLE
Material Derivative with Respect to x

### DISCUSSION THREAD

**Discussion by nsundquist** [idaholab/moose#18415] [Q&A General] [answered]
Hi everyone,

I'm currently trying to create a custom kernel and I need to calculate the derivative of a material with respect to x. I have been using SplitCHCParsed.C as an example to do this.

This is the code I have right now for getting the derivative:

`
CoefDiffusionMod::CoefDiffusionMod(const InputParameters & parameters)
    :  DerivativeMaterialInterface<JvarMapKernelInterface<SplitCHCRes>>(parameters),
       _dFdc(getMaterialPropertyDerivative<Real>("m_name", _var.name())) {
}
`

`
Real
CoefDiffusionMod::computeDFDC(PFFunctionType type) {
      return _dFdc[_qp];
}
`

I hope someone can shed some light on how to use x in place of _var.name() if it is possible. 

Thanks,

Nathan

**Comment by GiudGiud** [comment]
Hello

So you're looking for the first component of the gradient. You can't take gradients of material properties in moose right now, but you can use a MaterialRealAux kernel to turn the material property into an aux variable, then get the gradient of that. 

The only issue is that aux variables do not leverage automatic differentiation, so the Jacobian will no longer be exact just using AD. If this becomes a problem, you can remedy this by re-writing the computeJacobian routine to add the contribution of the gradient of the material property manually. 

Please do not paste screenshots in your future posts, they are not searchable

Guillaume

**Reply by GiudGiud** [reply]
if you are on a longer time scale (at least a few months), we are in the midst of developing a new functor-based on-demand material system that will be able to do this.

**Reply by GiudGiud** [reply]
Finally 3rd option, inspired by @hugary1995 's idea, you can define a material that is simply grad_F (or dF_dx if you only need the x-derivative) and do the chain rule yourself, using the gradient of each variable this material depends on.

**Comment by hugary1995** [comment, marked as answer]
@dschwen Do you think it is possible to let derivative parsed material to compute spatial gradient? I am just curious what if we ask the users to provide all coupled variables to `args`, and then use the chain rule to provide the gradients, like
```
dF_dx = dF_da * da_dx + dF_db * db_dx + dF_dc * dc_dx
```
where `args = 'a b c'`.

**Reply by hugary1995** [reply]
Urgh, I guess then we are ignoring all chain rules associated with inhomogeneous material properties...

**Reply by dschwen** [reply]
@hugary1995 DerivativeParsedMaterial has support for chain rule derivatives for inhomogeneous material properties!

**Reply by dschwen** [reply]
So in short, it is possible, but it will require a bit more work (as the inhomogeneous material properties require two layers of chain rule derivatives).

We're doing the gradient calculation this way routinely in the phase field module!
